Import changes from upstream prelude

avdv opened the pull request:
This PR introduces a few changes from upstream, mostly in order to reduce the amount of changes between the
current prelude and ourselves to make it easier to upstream, and to benefit from upstream changes.

- **Dedupe attribute lists of deps**
- **Use `at_argfile` helper**
- **Remove left-over comment**
- **Add performance optimisations for handling package DBs**
- **Move package db selection before the loop**
- **Use attr_deps_haskell_toolchain_libraries**
- **Remove `exposed_package_libs`**
- **Reformat**
- **Convert cxx toolchain LinkerType to enum**
- **Fix haskell_ghci() for multi-version GHC**
- **Rename relative args -> args**

ndmitchell committed:
Dedupe attribute lists of deps

Summary: If we duplicate a library twice then the GHC package file becomes an error. We can use tsets to dedupe between instances, but within a list of children, add dedupe here to remove the duplicates.

Reviewed By: shayne-fletcher

Differential Revision: D56241182

fbshipit-source-id: 0d5ae8f2f48eeb7945986a19d1778a4f2d217b21

ndmitchell committed:
Optimise generating ghci targets

Summary: D56378115 optimised 2 places we generate .db files. There was one remaining one which I missed - now caught.

Reviewed By: helfper

Differential Revision: D56642470

fbshipit-source-id: 83979f9ebdce0877929cb15b42bf39b0acd2de49

chatura-atapattu committed:
Rename relative args -> args

Summary: Because we had absolute args before, we needed to differentiate. We no longer need to do this. By default we assume relative args.

Reviewed By: lmvasquezg

Differential Revision: D56731576

fbshipit-source-id: 90bbc0021171d8640a5da900b78315a6f8271542
(cherry picked from commit f955fc2cf8e72b148d771f73e54a64cf6969e926)

Commit:
Fix haskell_ghci() for multi-version GHC

Summary:
Including tp2_build of ghc/8.8.3/platform010 and ghc/9.2.4/platform010

allow-large-files

Reviewed By: pepeiborra

Differential Revision: D61471626

fbshipit-source-id: fb211301cb0d29787e27bf4348e520916ddc9f1e
(cherry picked from commit 1848949d99189e417fbb64caa3009c6f7651797e)

Commit:
Convert cxx toolchain LinkerType to enum

Summary: Because I am interested in using this as a dict key in the next diff. `dict[LinkerType, RunInfo]`

Reviewed By: capickett

Differential Revision: D62385685

fbshipit-source-id: 2e47ece5cf5ca8180effa12e83021cd0ba73012a
(cherry picked from commit 446425115ec180ea6f7ee638e081dd968ac88090)

avdv committed:
Use `at_argfile` helper

avdv committed:
Remove left-over comment

avdv committed:
Add performance optimisations for handling package DBs

See commit 997574cb277b58f32c7b8e91a1333d49ee017d73:

multiple packages might exist within the same package_db. Before, since we deduplicated everything by value, you might have 100 packages
and 20 package_dbs. If you give GHC 100 package dbs (even if there are only 20 unique ones) the performance crashes dramatically.

That's especially true if you have a package_db representing all of stackage, which both occurs a lot and is slow to load. The two places are:

* When we construct GHC_PACKAGE_PATH env var. These are fairly trivially deduped with a dict construction
* When we do packagedb_args.add. We traverse the TSet in the loop above, so dedupe to a dict then too.

avdv committed:
Move package db selection before the loop

avdv committed:
Use attr_deps_haskell_toolchain_libraries

avdv committed:
Remove `exposed_package_libs`

They are always empty.

avdv committed:
Reformat

avdv committed:
Gracefully handle `None` values for ghci

aherrmann commented on haskell/compile.bzl:
```diff
@@ -356,7 +354,26 @@ def get_packages_info2(
+            lib.import_dirs.values(),
+            lib.stub_dirs,
+            # libs of dependencies might be needed at compile time if
+            # we're using Template Haskell:
```
Just to be sure, this is not used in our fine grained module compile actions, correct? We don't want to depend on libraries in that case.

avdv replied:
Yes, good catch. I was thinking about adding a flag to enable / disable this behavior to add backward compatibility for vanilla ghc but I want to tackle that separately. Removed.

\edit: Just removed the comment. The libs are no longer needed for TH, but are needed for the `exposed_package_args`.

aherrmann commented on haskell/compile.bzl:
```diff
@@ -356,7 +354,26 @@ def get_packages_info2(
+            lib.libs,
+        ])
+        exposed_package_args.add(hidden_args)
+        packagedb_args.add(hidden_args)
```
IIUC this is a change compared to before. Before the libraries were only added to `exposed_package_args`. Now they're also added to `packagedb_args`. Why is that necessary?

avdv replied:
Right, this is not needed. Removed.

aherrmann reviewed (approved):
Looks good, but some questions about the libraries handling. I'd like to make sure we're not inadvertently introducing dependencies on link actions to metadata or compile actions.

stepancheg committed:
Avoid list["xxx"] in prelude

Summary: Because string literals in types are deprecated and need to be removed.

Reviewed By: JakobDegen

Differential Revision: D63408317

fbshipit-source-id: 11552d44893aaef23e097dd9fda84fdaf35eb0cc
(cherry picked from commit 9297f2c298c1f0cd8fbbbb9c9fff31930881db90)

avdv committed:
Remove boot files from ghc command line

This fixes a regression introduced in commit de7f2c44275d88e1de528f0923e6140c0bfceaee which broke compilation
of sources with cyclic module dependencies.
